src/JSONPath/JSONObject.hack

```diff
@@ -0,0 +1,484 @@
+        $result = $this->getMatching($jsonObject, $jsonPath);
+        if (C\is_empty($result)) {
+            return null;
+        }
+
+        if ($this->smartGet && !$this->hasDiverged) {
+            return new WrappedResult($result[0]['object']);
+        }
+
+        return new WrappedResult(Vec\map($result, $r ==> $r['object']));
+    }
+
+    /**
+     * Replaces the element that results from the $jsonPath query
+     * to $value. This method disallows divering (wildcard) paths.
+     * It is a no-op if the given path doesn't already exist.
+     *
+     * This method returns a new JsonObject with the new value.
+     *
+     * @param string $jsonPath jsonPath
+     * @param mixed $value value
+     *
+     * @return mixed
+     */
+    public function replace(string $jsonPath, mixed $value): WrappedResult<JSONObject> {
+        if (Str\trim($jsonPath) === self::TOK_ROOT) {
+            return new WrappedResult(new JSONObject(\json_encode($value)));
+        }
+
+        $this->hasDiverged = false;
+        $result = $this->getMatching($this->jsonObject, $jsonPath);
+        if ($this->hasDiverged) {
+            throw new DivergentJSONPathSetException('Cannot set a value using a wildcard JSON path');
+        }
+
+        $out = self::setPathsToValue(
+            shape('object' => $this->jsonObject, 'path' => vec[]),
+            Vec\map($result, $p ==> $p['path']),
+            $value,
+        );
+
+        return new WrappedResult(new JSONObject(\json_encode($out)));
+    }
+
+    private static function pathMatched(vec<ExplodedPathType> $paths, ExplodedPathType $path): bool {
+        return C\contains($paths, $path);
+    }
+
+    private static function setPathsToValue(ObjectAtPath $object, vec<ExplodedPathType> $paths, mixed $value): mixed {
+        $jsonObject = $object['object'];
+        $path = $object['path'];
+
+        if ($jsonObject is vec<_> || $jsonObject is dict<_, _>) {
+            $out = $jsonObject is vec<_> ? vec[] : dict[];
+            foreach ($jsonObject as $key => $original_value) {
+                $childPath = Vec\concat($path, vec[$key]);
+                $newValue = self::pathMatched($paths, $childPath)
+                    ? $value
+                    : self::setPathsToValue(shape('object' => $original_value, 'path' => $childPath), $paths, $value);
+
+                if ($out is vec<_>) {
+                    $out[] = $newValue;
+                } else if ($out is dict<_, _>) {
+                    $out[$key] = $newValue;
+                }
+            }
+
+            return $out;
+        }
+
+        return $jsonObject;
+    }
+
+    private static function matchArrayIndex(string $jsonPath): ?shape('index' => string, 'rest' => string) {
+        if (Str\is_empty($jsonPath) || $jsonPath[0] != self::TOK_SELECTOR_BEGIN) {
+            return null;
+        }
+
+        $matches = null;
+        if (\preg_match_with_matches(self::RE_ARRAY_INDEX, $jsonPath, inout $matches)) {
```

Comment:
would it work to use HSL Regex here? (`HH\Lib\Regex\every_match()` I think?) We try to use HSL regex everywhere if possible - the only thing it doesn't support is dynamic regex strings but for these static regexes it should work and it provides a shape describing the match output including capture groups which helps keep things nice and safe (I also find it's nice to name the capture groups to make that easier to understand).

Reply:
Switched them over. Also learned in the process, that you can't use a `re"\\"` prefix string as a constant...